// pwi-plata-type/__BUILD__/libs/models/model-template-union.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelTemplateUnion = void 0;
class ModelTemplateUnion {
    config;
    templates;
    type;
    filters;
    constructor(config, templates) {
        this.config = config;
        this.templates = templates;
        this.filters = new Set();
    }
    addFilter(filter) {
        this.filters.add(filter);
    }
    getExtractFromUnion(value) {
        let type = `${value[this.config.key] ?? null}`;
        value[this.config.key] = value[this.config.key] ?? null;
        const template = this.templates[type];
        if (template === undefined) {
            return {
                errorID: 'PLMODUNEXTFRUN001',
                msg: `O Valor de ${this.config.key} tem que ser ${Object.keys(this.templates).join(',')}`,
                error: type
            };
        }
        return {
            model: template,
            value: value,
        };
    }
    filter(value) {
        const extraced = this.getExtractFromUnion(value);
        if (extraced.errorID !== undefined) {
            return value;
        }
        if (extraced.model.filter !== undefined) {
            value = extraced.model.filter(value);
        }
        this.filters.forEach(f => { value = f(value); });
        return value;
    }
    async validate(value, skipFilter) {
        const extraced = this.getExtractFromUnion(value);
        if (extraced.errorID !== undefined) {
            return {
                errors: [extraced],
                value: undefined
            };
        }
        if (!skipFilter) {
            this.filters.forEach(f => { value = f(value); });
        }
        const result = await extraced.model.validate(extraced.value, skipFilter);
        if (result.errors !== undefined) {
            return {
                errors: result.errors,
                value: undefined
            };
        }
        return {
            errors: undefined,
            value: result.value
        };
    }
}
exports.ModelTemplateUnion = ModelTemplateUnion;
