// pwi-plata-type/__BUILD__/libs/models/model-template-combine.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModelTemplateCombine = void 0;
class ModelTemplateCombine {
    type;
    models;
    validations;
    filters;
    constructor(models) {
        this.models = models;
        this.validations = new Set();
        this.filters = new Set();
    }
    addValidation(callback) {
        this.validations.add(callback);
    }
    addFilters(callback) {
        this.filters.add(callback);
    }
    filter(obj) {
        let value = obj;
        for (let i = 0; i < this.models.length; i++) {
            const model = this.models[i];
            if (model.filter !== undefined) {
                value = model.filter(value);
            }
        }
        this.filters.forEach(f => { value = f(value); });
        return value;
    }
    async validate(obj, skipFilter) {
        let v = obj;
        if (!skipFilter) {
            v = this.filter(v);
        }
        {
            const promises = [];
            for (let i = 0; i < this.models.length; i++) {
                const model = this.models[i];
                promises.push(Plata.FastPromise(() => model.validate(v, true)));
            }
            const results = await Promise.all(promises);
            const errors = results
                .filter((r) => r.errors !== undefined)
                .flatMap(r => r.errors);
            if (errors.length !== 0) {
                return {
                    value: undefined,
                    errors
                };
            }
        }
        {
            const promises = [];
            this.validations.forEach(va => promises.push(Plata.FastPromise(() => va(v))));
            const result = await Promise.all(promises);
            const errors = result.filter(Boolean);
            if (errors.length !== 0) {
                return {
                    errors,
                    value: undefined
                };
            }
        }
        return {
            value: v,
            errors: undefined
        };
    }
}
exports.ModelTemplateCombine = ModelTemplateCombine;
