// pwi-plata-type/__BUILD__/libs/axios.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlataAxios = PlataAxios;
const axios_1 = __importDefault(require("axios"));
async function PlataAxios(config) {
    return (0, axios_1.default)(config).then(resp => resp, err => {
        if (err.response) {
            return err.response;
        }
        if (err.request) {
            return {
                errorID: 'PLAX0001',
                msg: 'Erro ao se comunicar com API externa',
                error: {
                    message: err?.toString(),
                    stack: err?.stack ?? new Error().stack
                }
            };
        }
        return {
            errorID: 'PLAX0002',
            msg: 'Erro inesperado ao se comunicar com API externa',
            error: {
                message: err?.toString(),
                stack: err?.stack ?? new Error().stack
            }
        };
    });
}
