// pwi-plata-type/__BUILD__/bin/runtime/effect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlataRuntimeE = void 0;
const effect_1 = require("effect");
var PlataRuntimeE;
(function (PlataRuntimeE) {
    async function PromiseAll(promises) {
        if (promises.length === 0) {
            return [];
        }
        const results = [];
        let error = null;
        const promisesWithID = promises.map((p, i) => ({
            promise: p,
            id: i
        }));
        const effect = effect_1.Effect.forEach(promisesWithID, (p) => effect_1.Effect.gen(function* (_) {
            const result = yield* _(effect_1.Effect.promise(() => p.promise));
            if (result.errorID !== undefined) {
                error = result;
                return yield* _(effect_1.Effect.interrupt);
            }
            results[p.id] = result;
        }), { concurrency: 'unbounded' });
        await effect_1.Effect.runPromise(effect).catch((err) => {
            if (error === null)
                error = {
                    errorID: 'PBRUNEFFPROALL001',
                    msg: 'Erro ao realizar a operação',
                    error: {
                        message: err?.toString(),
                        stack: err?.stack ?? new Error().stack
                    }
                };
        });
        if (error !== null) {
            return error;
        }
        return results;
    }
    PlataRuntimeE.PromiseAll = PromiseAll;
})(PlataRuntimeE || (exports.PlataRuntimeE = PlataRuntimeE = {}));
