// pwi-plata-type/__BUILD__/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlataFs = exports.PlataSql = exports.PlataAxios = exports.PlataSwagger = exports.PlataModels = exports.PlataTools = exports.PlataRoutes = exports.PlataCluster = void 0;
const PlataCluster = __importStar(require("./libs/cluster"));
exports.PlataCluster = PlataCluster;
const PlataRoutes = __importStar(require("./libs/router"));
exports.PlataRoutes = PlataRoutes;
const PlataTools = __importStar(require("./libs/tools"));
exports.PlataTools = PlataTools;
const PlataModels = __importStar(require("./libs/models"));
exports.PlataModels = PlataModels;
const PlataSwagger = __importStar(require("./libs/swagger"));
exports.PlataSwagger = PlataSwagger;
const PlataSql = __importStar(require("./libs/sql"));
exports.PlataSql = PlataSql;
const fs_1 = require("./libs/fs");
Object.defineProperty(exports, "PlataFs", { enumerable: true, get: function () { return fs_1.PlataFs; } });
const axios_1 = require("./libs/axios");
Object.defineProperty(exports, "PlataAxios", { enumerable: true, get: function () { return axios_1.PlataAxios; } });
