// pwi-plata-type/src/libs/axios.ts
import axios, { AxiosResponse, AxiosRequestConfig } from 'axios'
import { PlataResultado } from './tools';

export async function PlataAxios<T = any>(config: AxiosRequestConfig): Promise<PlataResultado<AxiosResponse<T>>> {
    return axios(config).then(
        resp => resp,
        err => {
            if (err.response) {
                return err.response
            }

            if (err.request) {
                return {
                    errorID: 'PLAX0001',
                    msg: 'Erro ao se comunicar com API externa',
                    error: {
                        message: err?.toString(),
                        stack: err?.stack ?? new Error().stack
                    }
                }
            }

            return {
                errorID: 'PLAX0002',
                msg: 'Erro inesperado ao se comunicar com API externa',
                error: {
                    message: err?.toString(),
                    stack: err?.stack ?? new Error().stack
                }
            }
        }
    )
}
